### DISCUSSION TITLE
Gomponents vs Templ article : fairness check

### DISCUSSION THREAD

**Discussion by EwenQuim** [a-h/templ#948] [General]
Hi Adrian, 

I'm writing an article about Go templating, about **Gomponents vs Templ** and I wanted to make sure I have a state-of-the-art article about your library.

Is it possible to read it and tell me if I get your library wrong somewhere? Of course, I've sent the article to Gomponents team for proofreading too.

https://nuage.quimerch.com/-/ewen/articles/templ-vs-gomponents

Thanks again for your lib ❤️

**Comment by a-h** [comment]
In terms of getting the tools to work together. To use a gomponent in templ requires a couple of lines of code to adapt it.

```go
type Gomponent interface {
  Render(w io.Writer) error
}

func ToTempl(g Gomponent) templ.Component {
  func templ.ComponentFunc(func(ctx context.Context, w io.Writer) error {
     return g.Render(w)
  })
}
```

To go the other way requires a couple of extra lines.

```
type gomp struct {
   c templ.Component
}

func (g gomp) Render(w io.Writer) error {
  return g.c.Render(context.Background(), w)
}

func ToGomponent(c templ.Component) Gomponent {
  return g{c: c}
}
```

Would be cool if they interoperated directly.

> In 1-2 years, it might stable and more powerful that Gomponents, but I feel it’s not quite there yet.

Yeah, well, you know, that's just, like, your opinion, man. 😁

**Reply by EwenQuim** [reply]
Thank you very much for the feedback!

> Yeah, well, you know, that's just, like, your opinion, man. 😁

Haha thanks for taking it well, because you know, the site on which you just read the article is made with Templ ;)

**Comment by jhelberg** [comment]
It would be nice to publish some performance figure. I haven't done anyhting quantitative, but I see a drastic performance improvement in templ over html/template.

Also, I did not go into this yet, the templ-solution allows for a incremental (call it streaming) feed of html to the client-browser. I don't think either gomponent or html/template will ever be able to do this as creating the html-content is seperate from rendering it. I will do this some time, as my html-pages contain lots of database-fetched sensor-data.

The third thing you might want to address is the language cardinality. For html/template one needs to learn a new language. For Gomponent and templ one does not.

**Reply by EwenQuim** [reply]
Thanks for the feedback!

**Reply by jhelberg** [reply]
meanwhile. ive dont some performance measurement for a single html-page I'm making; I've done 4 reload of each:

    - templ :: 3.04ms, 3.86ms, 2.92ms, 2.88ms
      1700 lines of templ-code; 3500 chars of html

    - html/template :: 38.30ms, 41.60ms, 40.43ms, 47.55ms
      1062 lines of template; 3500 chars of html

I'm only timing the rendering. Huge difference! I expceted a factor of 5, not 10.

**Reply by EwenQuim** [reply]
Can you provide us the page and the source ? This can have a huge impact

**Comment by jhelberg** [comment]
and one more with regards to differences: both gomponent and templ need balanced html-tags within a code-block. html/template doesnt require that as it is mainly language-agnostic. In some situations, it is more effective to have opening and closing html-elements in their own code-blocks. This is only possible, withing non-raw context, in html/template. I find that very confusing, as there is only a hygienic argument for balanced elements, not a necessary one.
